Retry tuya setup on auth rate limiting

AlexSzlavik opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Use https://github.com/PaulAnnekov/tuyaha/pull/60 to detect and fix authentication rate limiting by tuya.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/core/issues/43840
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
Workaround for tuya auth rate limiting

Until https://github.com/PaulAnnekov/tuyaha/pull/60 or similar is merged to tuyaha, this workaround detects and fixes authentication rate limiting by tuya.

homeassistant commented:
Hi @AlexSzlavik, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44001).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Fix black error

homeassistant commented:
Hi @AlexSzlavik, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44001).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Flake8 fixes, squashme

ollo69 commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -120,29 +120,36 @@ async def async_setup(hass, config):
     """Set up Tuya platform."""
 
-    tuya = TuyaApi()
+    tuya = hass.data.get(DOMAIN, {}).get(TUYA_DATA)
```
Probably it is better to add a comment, as `#Try to get tuya object created in config flow`

Commit:
Add comment regarding re-use of connection

Commit:
Fixing trailing whitespace

bbaros commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -108,6 +108,7 @@ def _try_connect(self):
         except TuyaAPIException:
             return RESULT_AUTH_FAILED
 
+        self.hass.data[DOMAIN] = {TUYA_DATA: tuya}
```
How about doing this if tuya.init returns true?  If false, reset the store so that auth and connection will be started from scratch.

```
if tuya.init(...): 
    self.haas.data[DOMAIN] = {TUYA_DATA: tuya}
else:
    self.haas.data[DOMAIN] = {TUYA_DATA: None}
```

MartinHjelmare replied:
We don't allow the config flow to set instances in hass.data. The config flow should be isolated.

AlexSzlavik replied:
@MartinHjelmare sounds like this is a "best practice" restriction? The code did run when I tested it. I'm happy to change this around but I'm not sure if there is any other global state, or way to forward config objects to the actual setup of the component.

AlexSzlavik replied:
Maybe there is some other established pattern for handling heavy rate limiting like this already. Open to any suggestions.

ollo69 replied:
Now we have an integration that say OK during configuration and than is not created. If this solution is not applicable, probably is necessary to find an alternative that delay component creation and inform the user after integration setup.

MartinHjelmare replied:
When the library has been updated to raise the specific exception we can catch that in `async_setup_entry` and raise `ConfigEntryNotReady`. That will retry the setup.

ollo69 replied:
So we know that cannot be done. At this point only solution is wait for upstream PR and create new PR here when ready.

AlexSzlavik replied:
That's still a pretty miserable experience though. Even if we catch and wait, the UI will look as though everything froze. Hopefully tuya gets back to me and we can make a change to the API to better handle this. In the meantime I guess this PR can be declined.

MartinHjelmare replied:
The UI won't look frozen. This problem only happens when first creating the integration. So nothing from this integration will show up until the entry setup finishes after the retry is successful.

AlexSzlavik replied:
I'll try this again shortly, maybe I did something wrong before

darksession95 commented:
Sorry im not a coder, but does this bug prevent Tuya to communicate with Home Assistant? Because im trying to get Smartlife/Tuya to work and nothing shows up.

AlexSzlavik commented:
> Sorry im not a coder, but does this bug prevent Tuya to communicate with Home Assistant? Because im trying to get Smartlife/Tuya to work and nothing shows up.

After you add tuya , restart ha core and things should work :)

AlexSzlavik commented:
I'll try this again shortly, maybe I did something wrong before

On Mon, Dec 7, 2020 at 9:34 AM Martin Hjelmare <notifications@github.com>
wrote:

> *@MartinHjelmare* commented on this pull request.
> ------------------------------
>
> In homeassistant/components/tuya/config_flow.py
> <https://github.com/home-assistant/core/pull/44001#discussion_r537553664>:
>
> > @@ -108,6 +108,7 @@ def _try_connect(self):
>          except TuyaAPIException:
>              return RESULT_AUTH_FAILED
>
> +        self.hass.data[DOMAIN] = {TUYA_DATA: tuya}
>
> The UI won't look frozen. This problem only happens when first creating
> the integration. So nothing from this integration will show up until the
> entry setup finishes after the retry is successful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/44001#discussion_r537553664>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIKXNFGL4VZXOQAWKUXHSTSTTRY5ANCNFSM4UPUJD3Q>
> .
>

Commit:
Workaround for tuya auth rate limiting

Until https://github.com/PaulAnnekov/tuyaha/pull/60 or similar is merged to tuyaha, this workaround detects and fixes authentication rate limiting by tuya.

AlexSzlavik commented:
@MartinHjelmare @ollo69 I've collapsed the temporary fix PR to just detect the rate limiting case and retry.
After locally testing, the behaviour is (assuming no login attempt within 60 seconds):
1. Add Tuya integration, succeeds
2. The integration will attempt to fetch devices, but fail while trying to connect to Tuya, logs:
```2020-12-07 09:42:18 ERROR (MainThread) [homeassistant.components.tuya] Tuya login ratelimited
2020-12-07 09:42:18 WARNING (MainThread) [homeassistant.config_entries] Config entry for tuya not ready yet. Retrying in 5 seconds
2020-12-07 09:42:24 ERROR (MainThread) [homeassistant.components.tuya] Tuya login ratelimited
2020-12-07 09:42:24 WARNING (MainThread) [homeassistant.config_entries] Config entry for tuya not ready yet. Retrying in 10 seconds
2020-12-07 09:42:34 ERROR (MainThread) [homeassistant.components.tuya] Tuya login ratelimited
2020-12-07 09:42:34 WARNING (MainThread) [homeassistant.config_entries] Config entry for tuya not ready yet. Retrying in 20 seconds
2020-12-07 09:42:55 ERROR (MainThread) [homeassistant.components.tuya] Tuya login ratelimited
2020-12-07 09:42:55 WARNING (MainThread) [homeassistant.config_entries] Config entry for tuya not ready yet. Retrying in 40 seconds
2020-12-07 09:43:35 INFO (MainThread) [homeassistant.components.tuya] Tuya discovery device poll interval set to 605 seconds
2020-12-07 09:43:35 INFO (MainThread) [homeassistant.components.tuya] Tuya query device poll interval set to 120 seconds
2020-12-07 09:43:35 INFO (MainThread) [homeassistant.components.switch] Setting up switch.tuya
2020-12-07 09:43:35 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.tuya entity: switch.08282601840d8e7a1cd8
2020-12-07 09:43:35 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new switch.tuya entity: switch.10051061600194dcff96
```
3. After retrying for more than 60seconds, the integration will login succesfully and fetch the devices. Now everything works

Caviats: While the integration is attempting to get the devices, if a user clicks on the "rename" or "options" button of the integration, a "Wait until the integration is installed" spinner shows up. This spinner can only be dismissed by refresh the page. I think it's worth documenting this behaviour. Thoughts? 
Also PTAL at this minimal PR now.

ollo69 commented:
It's ok for me, but I don't know if it is acceptable to use the error text or we have to wait for new TuyaHA version. May be @PaulAnnekov can help us merging [PR #60](https://github.com/PaulAnnekov/tuyaha/pull/60).

MartinHjelmare commented:
The caveat behavior should probably be improved upon in the frontend to handle better.

AlexSzlavik commented:
> It's ok for me, but I don't know if it is acceptable to use the error text or we have to wait for new TuyaHA version. May be @PaulAnnekov can help us merging [PR #60](https://github.com/PaulAnnekov/tuyaha/pull/60).

Note that this is the same check that we'd be making in the tuyaha library. It's not great to check for the specific string and highly brittle, but it's all we got to work with. Hopefully Tuya will respond to my requsets.

> The caveat behavior should probably be improved upon in the frontend to handle better.
SGTM, I'll look into whether this can be improved as a follow up

MartinHjelmare commented:
We should also guard in the options flow against a config entry that is not set up, and abort the options flow in that case. This should solve the KeyError in the linked issue.

https://github.com/home-assistant/core/blob/70133f2096984610cc916e54a1158d9417a10051/homeassistant/config_entries.py#L56-L57

Commit:
Abort Options flow if tuya api unavailable

AlexSzlavik commented:
> We should also guard in the options flow against a config entry that is not set up, and abort the options flow in that case. This should solve the KeyError in the linked issue.
> 
> https://github.com/home-assistant/core/blob/70133f2096984610cc916e54a1158d9417a10051/homeassistant/config_entries.py#L56-L57

Added the requested abort and tested locally. Much better behaviour now, seeing an abort message instead of just hanging. PTAL

PaulAnnekov commented:
Merged https://github.com/PaulAnnekov/tuyaha/pull/60 and released new version https://github.com/PaulAnnekov/tuyaha/releases/tag/0.0.9.

MartinHjelmare commented:
@AlexSzlavik Please bump `tuyaha` to 0.0.9 here and use the new exception. Then we can merge I think.

Commit:
Bump tuyaha to 0.0.9 and use new api exception

Commit:
Expose new auth exception to config flow

ollo69 commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -137,6 +138,10 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     ) as exc:
         raise ConfigEntryNotReady() from exc
 
+    except TuyaAPIRateLimitException as exc:
+        _LOGGER.error("Tuya login ratelimited")
+        raise ConfigEntryNotReady() from exc
+
```
I think that this can be grouped with others exceptions and you don't need the log, you already have `from exc`

AlexSzlavik replied:
I personal like the log, but am open to removing it. The current error codes (and translations) don't really tell the user that this is a transient error. The logs make this more clear

AlexSzlavik replied:
oh, I miss-read what you mean. The `TuyaAPIRateLimitException` is stacked with `ConfigEntryNotReady`. Let me see if that gets unwrapped in logs.

AlexSzlavik replied:
hrm, config_entries does not print out the actual exception, or stack. I'd vote for keeping the log, but if you want to remove it I'll change it.

ollo69 replied:
If you want to leave it in my opinion should be a warning, but let @MartinHjelmare suggest best solution based on best practice.

MartinHjelmare replied:
It's good to keep the log.

AlexSzlavik commented:
Thanks @ollo69 @MartinHjelmare for your patience on my first PR to HA. ❤️

MartinHjelmare commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -137,12 +138,11 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
-            _LOGGER.error("Tuya login ratelimited")
-            raise ConfigEntryNotReady() from exc
+    except TuyaAPIRateLimitException as exc:
+        _LOGGER.error("Tuya login ratelimited")
```
```suggestion
        _LOGGER.error("Tuya login rate limited")
```

MartinHjelmare committed:
Adjust words

MartinHjelmare reviewed (approved):
Thanks!
